### ISSUE TITLE
Evaluate alternatives to rendering dbt projects other than during DAG parsing time

### ISSUE THREAD

**Issue by tatiana** [astronomer/astronomer-cosmos#520]
At the moment, Cosmos does several things during DAG parsing time:
- It may convert [connection -> profile](https://astronomer.github.io/astronomer-cosmos/profiles/index.html)
- It [parses the dbt project](https://astronomer.github.io/astronomer-cosmos/configuration/parsing-methods.html), to convert it into an Airflow DAG (dynamically, during parse-time)
- It may apply filters based on user-defined selectors if using `LoadMode.DBT_MANIFEST` or `LoadMode.CUSTOM`

We have yet to do a performance analysis (#519). However, this could bring issues to the scheduler in larger dbt projects.

Ideally, we should avoid this logic during scheduling / DAG parsing. Some alternatives:
1. Allow users to pre-compute an expanded Airflow DAG
2. Use [Airflow dynamic task mapping](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/dynamic-task-mapping.html)
3. Before running Airflow, extend manifest definitions to include filtered nodes (materialising into a JSON) before DAG-parsing
4. See if we could leverage [DAG setup and teardown](https://airflow.apache.org/docs/apache-airflow/stable/howto/setup-and-teardown.html) in any way
5. Others..?

Thoughts and ideas are very welcome!

**Comment by DanMawdsleyBA** [comment]
I would welcome this! One thing that might be useful is if I wanted to run the airflow dag with a different select statement. For example I only want to rerun part of the dag not the entire dag. If my process loads from 10 sources in parallel but only 1 source file has come through it can just run the downstream of that one file. Could cutdown airflow dag times and unnecessary queries on the database.

**Comment by tatiana** [comment]
#airflow-dbt users reported having to increase `AIRFLOW__CORE__DAGBAG_IMPORT_TIMEOUT` to be able to parse DAGs with Cosmos:
https://apache-airflow.slack.com/archives/C059CC42E9W/p1696435273519979

**Comment by benjamin-awd** [comment]
FWIW switching from the default parsing mode to `LoadMode.DBT_MANIFEST` seems to have had a dramatic impact on DAG file processing durations -- I think this may be tangentially related to the DagBag import timeout issue.
![Screenshot 2023-10-06 at 11 33 13 AM](https://github.com/astronomer/astronomer-cosmos/assets/62495124/4f434c09-fdc6-4590-b439-acd5756daebb)

**Comment by jlaneve** [comment]
Jotting down some ideas here in hopes to inspire more conversations and thoughts about this!

## Pre-computing DAG structure

The idea here is to offload the compute/time taken during parse time and shift it to build time. I've seen it done 2 ways:

1. Use dbt to "pre-compute" the DAG by generating a manifest file
2. Use jinja templating to construct an actual dagfile based on a dbt manifest / dbt ls command

Cosmos currently supports building the DAG from a manifest file, but it assumes a user can construct the manifest themselves. There may be ways to extend Cosmos to make this easier to do.

Jinja templating is another interesting way of doing things, and very different than what Cosmos does today. It does feel like the most efficient, and Cosmos does already have most of the operators necessary.

Note that both these options require introducing a build step to either generate a manifest or template out a dagfile.

I consider the "Before running Airflow, extend manifest definitions to include filtered nodes (materialising into a JSON) before DAG-parsing" part of this, too - whether we run a build step against a full manifest/project or subset of the manifest/project, it's still a build step.


## Dynamic task mapping

This is tricky to imagine, but I'll give it an attempt... I can think of a few setups, but none of them seem to work:

1. Treat an entire dbt project / Cosmos TaskGroup as a mapped task. This wouldn't work because there are no dependencies within mapped tasks.
2. Try to layer the project into a series of mapped tasks. This doesn't solve the problem, because we'd still need to parse to determine the layers.
3. Try to layer the project into a set of mapped tasks broken out by node type (i.e. sources, models). This doesn't work well because there's no "eager" execution. A DAG would have to wait until all the sources are done before executing any of the models.

Can't think of a way to get this working, but open to ideas.


## Leverage [DAG setup and teardown](https://airflow.apache.org/docs/apache-airflow/stable/howto/setup-and-teardown.html)

This would require us to mutate the DAG structure mid-execution, assuming we want to use a setup task to generate the graph. I don't think DAGs can mutate their structure mid-execution, but I could be wrong here. Correct me if I'm wrong!


## Cache the `dbt ls` command (NEW)

We don't do _any_ caching right now, meaning every time the DAG is parsed, we're rerunning a full `dbt ls` command (potentially with `dbt deps`). What we could/should do instead is cache the output of `dbt ls` on the scheduler. The first time the DAG is parsed, we write the output to some path (something like `/cosmos/dag_1/partial_manifest.json`). Then, on subsequent parses, we can:

1. Check if `/cosmos/dag_1/partial_manifest.json` exists. If so, we need to make sure it's not "stale"
2. Check the actual dbt project to ensure there haven't been updates since the last `partial_manifest.json` construction. If there have, update `/cosmos/dag_1/partial_manifest.json`

Tricky thing here is the initial parse. Users are already facing issues with the DAG import timeout, so this doesn't really solve the initial problem (although it does _significantly_ decrease load on the scheduler).

Separately, this also doesn't work well across workers. If the only place we cache the manifest is on the scheduler, we fix DAG parsing issues. What we don't fix is the dbt command startup time per model execution; each time a model is executed, dbt's going to have to generate the full manifest.


## Run a separate component

Something I've been thinking about a lot recently is running a separate server/pod/API for dbt and Airflow. This stems from three issues that don't have straightforward solutions:
- DAG parse times / worker manifest creation times. It's tough to coordinate across all Airflow components
- Supporting async execution. Essentially all dbt compute happens in the database itself, but dbt requires a bash command to be held open for the duration of execution; because of this, it's tricky to support async operators
- Separate dbt and Airflow repos. When running Airflow + dbt at scale, generally you're going to have separate repos and consolidating those can be tricky (git submodules, cross-repo CI/CD, etc)

A potential solution here is to run a simple API server that can be responsible for a few things:
- accepting dbt projects as deployments; you don't deploy dbt to airflow, you deploy to this service
- constructing the graph / parsing the project
- spawning new subprocesses to hold open `dbt run` commands
- reporting status & logs

Then, the UX can be as follows:
- user deploys dbt code to this service via CI/CD in a dedicated repo (or subfolder of a monorepo)
- Cosmos hits an endpoint (something like `/my_project/graph?select=...`) to get a filtered view of the graph Airflow needs to construct; Cosmos can also cache this locally in case there are disruptions in the service
- when a Cosmos DAG is run, instead of running locally within Airflow, it hits an endpoint (something like `/my_project/run?model=...`) to start the run and defers itself
- the deferred operator runs in the triggerer and hits a status/logs endpoint to poll for updates

I think this is a pretty great UX and doesn't feel too difficult to get a prototype of.


--

Would love other folks to chime in with other ideas and feedback!

**Comment by pgoslatara** [comment]
@jlaneve Two responses to your ideas:

### Dynamic Task Mapping

Indeed, the lack of dependencies between mapped tasks as well as individualised downstream dependencies (i.e. independent task_2 depends on mapped task task_1a) is a big limitation. I do see mapped tasks possibly being useful in two scenarios:
* Seeds: If a user wants to generate all seeds (or a selection) then there are by definition no dependencies between these tasks.
* Tests: running `dbt test` by default runs a series of tests with no inter-dependencies, again an option for mapped tasks.

### DAG mutation

I've no experience doing this but Monzo's talk at Airflow Summit (available [here](https://www.youtube.com/watch?v=1b6Uu4M0ExY)) implies that this is what they do. And it's also dbt related so maybe there's a few ideas in there for Cosmos!

**Comment by dwreeves** [comment]
@jlaneve RE:

> Separate dbt and Airflow repos. When running Airflow + dbt at scale, generally you're going to have separate repos and consolidating those can be tricky (git submodules, cross-repo CI/CD, etc)

and:

> accepting dbt projects as deployments; you don't deploy dbt to airflow, you deploy to this service

Something on the simpler side that requires less infrastructure would be to support cloud storage for dbt projects + manifests.

```python

from cosmos import DbtDag, ProjectConfig

my_cosmos_dag = DbtDag(
    project_config=ProjectConfig(
        dbt_project_path="s3://dbt-project-in-cloud-storage/dbt_project/",
        manifest_path="s3://dbt-project-in-cloud-storage/dbt_project/target/manifest.json"
        conn_id="aws_default"
    ),
    # ...,
)
```

^ This does not address issues relating to async execution, but it does support the decentralized repo pattern.

Huge, huge caveat: this gets messy with execution out of the tmp dir: it walks back the benefits of the existing symlink approach, since you'd need to copy the whole directory from cloud storage to make it work.
